src/System.Threading.ThreadPool/tests/ThreadPoolTests.netcoreapp.cs

```diff
@@ -10,82 +11,120 @@ namespace System.Threading.ThreadPools.Tests
-        [InlineData(true)]
-        public async Task QueueUserWorkItem_PreferLocal_ValueTypeStateObjectPassedThrough(bool preferLocal)
+        [MemberData(nameof(TwoBools))]
+        public async Task QueueUserWorkItem_PreferLocal_ValueTypeStateObjectPassedThrough(bool preferLocal, bool useUnsafe)
         {
             var tcs = new TaskCompletionSource<int>();
-            ThreadPool.QueueUserWorkItem(s => s.tcs.SetResult(s.value), (tcs, value: 42), preferLocal);
+            if (useUnsafe)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(s => s.tcs.SetResult(s.value), (tcs, value: 42), preferLocal);
+            }
+            else
+            {
+                ThreadPool.QueueUserWorkItem(s => s.tcs.SetResult(s.value), (tcs, value: 42), preferLocal);
+            }
             Assert.Equal(42, await tcs.Task);
         }
 
         [Theory]
-        [InlineData(false)]
-        [InlineData(true)]
-        public async Task QueueUserWorkItem_PreferLocal_RunsAsynchronously(bool preferLocal)
+        [MemberData(nameof(TwoBools))]
+        public async Task QueueUserWorkItem_PreferLocal_RunsAsynchronously(bool preferLocal, bool useUnsafe)
         {
             await Task.Factory.StartNew(() =>
             {
                 int origThread = Environment.CurrentManagedThreadId;
                 var tcs = new TaskCompletionSource<int>();
-                ThreadPool.QueueUserWorkItem(s => s.SetResult(Environment.CurrentManagedThreadId), tcs, preferLocal);
+                if (useUnsafe)
+                {
+                    ThreadPool.UnsafeQueueUserWorkItem(s => s.SetResult(Environment.CurrentManagedThreadId), tcs, preferLocal);
+                }
+                else
+                {
+                    ThreadPool.QueueUserWorkItem(s => s.SetResult(Environment.CurrentManagedThreadId), tcs, preferLocal);
+                }
                 Assert.NotEqual(origThread, tcs.Task.GetAwaiter().GetResult());
             }, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default);
         }
 
         [Theory]
-        [InlineData(false)]
-        [InlineData(true)]
-        public async Task QueueUserWorkItem_PreferLocal_ExecutionContextFlowed(bool preferLocal)
+        [MemberData(nameof(TwoBools))]
+        public async Task QueueUserWorkItem_PreferLocal_ExecutionContextFlowedIfSafe(bool preferLocal, bool useUnsafe)
         {
             var tcs = new TaskCompletionSource<int>();
             var asyncLocal = new AsyncLocal<int>() { Value = 42 };
-            ThreadPool.QueueUserWorkItem(s => s.SetResult(asyncLocal.Value), tcs, preferLocal);
+            if (useUnsafe)
+            {
+                ThreadPool.UnsafeQueueUserWorkItem(s => s.SetResult(asyncLocal.Value), tcs, preferLocal);
+            }
+            else
+            {
+                ThreadPool.QueueUserWorkItem(s => s.SetResult(asyncLocal.Value), tcs, preferLocal);
+            }
             asyncLocal.Value = 0;
```

Comment:
>asyncLocal.Value = 0; [](start = 12, length = 21)

Is it possible the queued work item can finish executing before executing this line? if so the following check can fail at that time in case of useunsafe

Reply:
Right.  There should not be a race condition here.